doc/sysdbd.1.txt
sysdbd(1)
=========
:doctype: manpage

NAME
----
sysdbd - system management and inventory collection service

SYNOPSIS
--------
*sysdbd* ['options']

DESCRIPTION
-----------
*sysdbd* is a multi-backend system management and inventory collection daemon.
It stores system and inventory information about hardware and software
systems. This information is (continuously) collected from various
configurable backends.

The main daemon itself is the central instance managing all collected
information and doing the correlation of objects provided by different
backends. It basically acts as a database server. All data retrieval, any
further processing, storing and exporting of data is done by plugins.

OPTIONS
-------
*sysdbd* accepts the following command-line options.

*-C* '<file>'::
	The main configuration file. This file defines the behavior of *sysdbd* by
	specifying default settings and the plugins to be loaded.

*-D*::
	Do not daemonize on startup: Rather than starting *sysdbd* as a background
	process detached from the current terminal and session (the default), run
	the process in the foreground. This allows easy monitoring of *sysdbd*.

*-h*::
	Display a usage and help summary and exit.

*-V*::
	Display the version number and copyright information.

CLIENT AUTHENTICATION
---------------------

include::sysdb-auth.txt[]

FILES
-----

*/etc/sysdb/sysdbd.conf*::
	The default main configuration file for the SysDB daemon.

*/etc/sysdb/ssl/key.pem*::
*/etc/sysdb/ssl/cert.pem*::
	The default SSL private key and certificate files to use for SSL
	connections.

SIGNALS
-------

*sysdbd* accepts and handles the following signals:

*SIGINT*, *SIGTERM*::
	These signals initiate a clean shutdown and termination of the daemon.

*SIGHUP*::
	This signal causes the daemon to re-read the configuration and reconfigure
	all plugins. Plugins which are no longer in use will be disabled and
	unloaded. Open client connections are not affected by this.

EXIT CODES
----------
*0*::
	Success.

*1*::
	Failure (syntax or usage error).

BUGS
----
None known.

SEE ALSO
--------
manpage:sysdbd.conf[5]

The SysDB homepage: https://sysdb.io/

AUTHOR
------
SysDB was written by Sebastian "tokkee" Harl <sh@tokkee.org>.

COPYRIGHT
---------
Copyright (C) 2012-2014 Sebastian "tokkee" Harl <sh@tokkee.org>

This is free software under the terms of the BSD license, see the source for
copying conditions. There is NO WARRANTY; not even for MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.

// vim: set tw=78 sw=4 ts=4 noexpandtab spell spelllang=en_us :



ReleaseNotes
                             SysDB Release Notes
                            =====================

  The release notes contain noteworthy changes in each System DataBase (SysDB)
  release, with a summary of the most important changes at the top. A complete
  and very detailed list of changes can be obtained from the Git logs for each
  release.

 Compatibility Note:
 ~~~~~~~~~~~~~~~~~~~

  All pre 1.0 development snapshots are not considered stable yet. That is,
  all interfaces, including the plugin API, the network protocol, or the query
  language, may change in an incompatible way at any time. All changes are
  documented here.

 About SysDB:
 ~~~~~~~~~~~~
  “System DataBase” (SysDB) is a multi-backend system management and inventory
  collection service. It stores system and inventory information about
  hardware and software systems. This information is (continuously) collected
  from various configurable backends (inventory services, monitoring services,
  etc.) and stored in a graph-like hierarchy of generic objects. The store may
  be queried through a generic interface independent of the active backends.
  Object names are canonicalized before they are added to the store to ensure
  a consistent view of your infrastructure.

  <https://sysdb.io/>

Version 0.6.0, Release Date: 2014-11-17
---------------------------------------

  This release focuses on a clearer query language. The underlying
  architecture has been refactored to allow for more powerful and generic
  expressions.

   * core:
     - Added support for arrays. Elements may be of any other supported
       data-type but only integers, decimals, and strings are fully supported.
   * store:
     - Refactored matchers: all matchers now operate on one or two expressions
       or other matchers, replacing the old and less flexible conditional
       matchers.
     - Attribute values may now be used anywhere in an expression.
     - On data-type mismatch when comparing attribute values, the values will
       be cast to string (previously only done for regex matches).
     - Added support for querying an object's backends (array of strings).
     - Added support for iterating child objects and array values.
     - All matchers now return false if any operand is NULL (except for IS
       NULL).
     - All operations now return NULL if any operand is NULL.
     - Fixed behavior of not-operators (!=, !~) and let them not match if any
       of the operands is NULL.
     - Add support for looking up all object types (hosts, services, metrics).
     - Refactored the JSON formatter to be more flexible and powerful.
     - Record parent objects for each stored object.
   * frontend:
     - Added a separate parser mode for expressions.
     - Added a new queryable field 'name' to be used in place of the rather
       magical keyword 'host' when looking up hosts.
     - Improved error reporting in the parser and the (newly introduced)
       analyzer.
     - Apply filters to hosts in the FETCH command.
   * query language:
     - All operators may now be used in all places if the involved data-types
       support the respective operation. For example, regular expressions may
       be used to match any value. Non-string values are cast to strings for
       that purpose. IS NULL / IS NOT NULL may be applied to arbitrary
       expressions.
     - Introduced the 'IN' operator to check if a value or an array of values
       is included in an array.
     - Added support for arrays: [<elem1>,<elem2>,...]
     - Added support for string and array concatenation using the '||'
       operator.
     - Attribute names have to be strings instead of identifiers now.
     - Access object fields using '<field>' rather than '.<field>'.
     - Let the LOOKUP and FETCH commands support services and metrics.
     - Introduced 'FETCH service|metric <host>.<name' for fetching services
       and metrics.
     - Added support for 'ANY' and 'ALL' operators which apply a matcher to
       each value emitted by an iterator (a host's services and metrics or
       arrays). 'MATCHING ANY service|metric|attribute <cmp> <expr>' replaces
       'MATCHING service|metric|attribute <cmp> <value>' expressions.

Version 0.5.0, Release Date: 2014-10-06
---------------------------------------

  This release focuses on simplifications and more flexibility in the frontend
  and the query language. Most of those changes are not backward-compatible
  but easy client implementations.

   * sysdb:
     - Include priority when printing log messages.
     - Fixed handling of empty queries.
   * store:
     - When serializing the entire store (LIST command), return an array of
       host objects (rather than wrapping it into another object).
     - Fixed JSON format of time-series data.
     - Fixed error checks of invalid compare expressions and reject them.
     - Fixed lookups by back-end name.
     - Added support for comparing attributes with different types by
       comparing their string values.
   * frontend:
     - Clarified details about asynchronous messages and authentication
       options and made client implementation more robust.
     - Introduced a new message type encoding the JSON serialized response of
       a query and its data type.
     - Include the priority in log messages.
     - Include object types in FETCH, LIST, and LOOKUP messages.
     - Added support for including services and metrics in LIST responses and
       skip hosts without the respective children.
   * query language:
     - Added support for services and metrics to LIST command.
     - Changed syntax for attribute lookup to 'attributes[<name>]'.
     - Changed syntax for accessing queryable fields to '.<field>'.
   * utils:
     - strbuf: Improved memory management.

Version 0.4.0, Release Date: 2014-09-01
---------------------------------------

  This release features support for metrics and transparent access to a

src/daemon/sysdbd.c
/*
 * SysDB - src/sysdbd.c
 * Copyright (C) 2012 Sebastian 'tokkee' Harl <sh@tokkee.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#if HAVE_CONFIG_H
#	include "config.h"
#endif /* HAVE_CONFIG_H */

#include "sysdb.h"
#include "core/plugin.h"
#include "core/store.h"
#include "utils/error.h"

#include "frontend/sock.h"

#include "daemon/config.h"

#if HAVE_LIBGEN_H
#	include <libgen.h>
#else /* HAVE_LIBGEN_H */
#	define basename(path) (path)
#endif /* ! HAVE_LIBGEN_H */

#include <errno.h>

#include <sys/stat.h>
#include <fcntl.h>

#include <signal.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <unistd.h>

#include <pthread.h>

#ifndef CONFIGFILE
#	define CONFIGFILE SYSCONFDIR"/sysdb/sysdbd.conf"
#endif

#ifndef DEFAULT_SOCKET
#	define DEFAULT_SOCKET "unix:"LOCALSTATEDIR"/run/sysdbd.sock"
#endif

static sdb_plugin_loop_t plugin_main_loop = SDB_PLUGIN_LOOP_INIT;
static sdb_fe_loop_t frontend_main_loop = SDB_FE_LOOP_INIT;

static char *default_listen_addresses[] = {
	DEFAULT_SOCKET,
};

static void
sigintterm_handler(int __attribute__((unused)) signo)
{
	frontend_main_loop.do_loop = 0;
} /* sigintterm_handler */

static void
exit_usage(char *name, int status)
{
	printf(
"Usage: %s <options>\n"

"\nOptions:\n"
"  -C FILE   the main configuration file\n"
"            default: "CONFIGFILE"\n"
"  -D        do not run in background (daemonize)\n"
"\n"
"  -h        display this help and exit\n"
"  -V        display the version number and copyright\n"

"\nSysDB daemon "SDB_VERSION_STRING SDB_VERSION_EXTRA", "PACKAGE_URL"\n",
basename(name));
	exit(status);
} /* exit_usage */

static void
exit_version(void)
{
	printf("SysDBd version "SDB_VERSION_STRING SDB_VERSION_EXTRA", "
			"built "BUILD_DATE"\n"
			"using libsysdb version %s%s\n"
			"Copyright (C) 2012-2013 "PACKAGE_MAINTAINER"\n"

			"\nThis is free software under the terms of the BSD license, see "
			"the source for\ncopying conditions. There is NO WARRANTY; not "
			"even for MERCHANTABILITY or\nFITNESS FOR A PARTICULAR "
			"PURPOSE.\n", sdb_version_string(), sdb_version_extra());
	exit(0);
} /* exit_version */

static int
daemonize(void)
{
	pid_t pid;

	if ((pid = fork()) < 0) {
		char errbuf[1024];
		sdb_log(SDB_LOG_ERR, "Failed to fork to background: %s",
				sdb_strerror(errno, errbuf, sizeof(errbuf)));
		return errno;
	}
	else if (pid != 0) {
		/* parent */
		exit(0);
	}

	if (chdir("/")) {
		char errbuf[1024];
		sdb_log(SDB_LOG_ERR, "Failed to change working directory to "
				"the root directory: %s",
				sdb_strerror(errno, errbuf, sizeof(errbuf)));
		return errno;
	}

	/* detach from session */
	setsid();

	close(0);
	if (open("/dev/null", O_RDWR)) {
		char errbuf[1024];
		sdb_log(SDB_LOG_ERR, "Failed to connect stdin to '/dev/null': %s",
				sdb_strerror(errno, errbuf, sizeof(errbuf)));
		return errno;
	}

	close(1);
	if (dup(0) != 1) {
		char errbuf[1024];
		sdb_log(SDB_LOG_ERR, "Could not connect stdout to '/dev/null': %s",
				sdb_strerror(errno, errbuf, sizeof(errbuf)));
		return errno;
	}

	close(2);
	if (dup(0) != 2) {
		char errbuf[1024];
		sdb_log(SDB_LOG_ERR, "Could not connect stderr to '/dev/null': %s",
				sdb_strerror(errno, errbuf, sizeof(errbuf)));
		return errno;
	}
	return 0;
} /* daemonize */

static void *
backend_handler(void __attribute__((unused)) *data)
{
	sdb_plugin_collector_loop(&plugin_main_loop);
	sdb_log(SDB_LOG_INFO, "Shutting down backend thread");
	return NULL;
} /* backend_handler */

int
main(int argc, char **argv)
{
	char *config_filename = NULL;
	_Bool do_daemonize = 1;

	pthread_t backend_thread;

	struct sigaction sa_intterm;
	int status;

	sdb_error_set_logger(sdb_plugin_log);

	while (42) {
		int opt = getopt(argc, argv, "C:DhV");

		if (-1 == opt)
			break;

		switch (opt) {
			case 'C':
				config_filename = optarg;
				break;
			case 'D':
				do_daemonize = 0;
				break;

			case 'h':
				exit_usage(argv[0], 0);
				break;
			case 'V':
				exit_version();
				break;
			default:
				exit_usage(argv[0], 1);
		}
	}

	if (optind < argc)
		exit_usage(argv[0], 1);

	if (! config_filename)
		config_filename = CONFIGFILE;

	if ((status = daemon_parse_config(config_filename))) {
		if (status > 0)
			sdb_log(SDB_LOG_ERR, "Failed to parse configuration file.");
		else
			sdb_log(SDB_LOG_ERR, "Failed to load configuration file.\n"
					"\tCheck other error messages for details.");
		exit(1);
	}

	if (! listen_addresses) {
		listen_addresses = default_listen_addresses;
		listen_addresses_num = SDB_STATIC_ARRAY_LEN(default_listen_addresses);
	}

	memset(&sa_intterm, 0, sizeof(sa_intterm));
	sa_intterm.sa_handler = sigintterm_handler;
	sa_intterm.sa_flags = 0;

	if (sigaction(SIGINT, &sa_intterm, /* old action */ NULL)) {
		char errbuf[1024];
		sdb_log(SDB_LOG_ERR, "Failed to install signal handler for "
				"SIGINT: %s", sdb_strerror(errno, errbuf, sizeof(errbuf)));
		exit(1);
	}
	if (sigaction(SIGTERM, &sa_intterm, /* old action */ NULL)) {
		char errbuf[1024];
		sdb_log(SDB_LOG_ERR, "Failed to install signal handler for "
				"SIGTERM: %s", sdb_strerror(errno, errbuf, sizeof(errbuf)));
		exit(1);
	}

	if (do_daemonize)
		if (daemonize())
			exit(1);

	sdb_log(SDB_LOG_INFO, "SysDB daemon "SDB_VERSION_STRING
			SDB_VERSION_EXTRA " (pid %i) initialized successfully",
			(int)getpid());

	sdb_plugin_init_all();
	plugin_main_loop.default_interval = SECS_TO_SDB_TIME(60);

	memset(&backend_thread, 0, sizeof(backend_thread));
	if (pthread_create(&backend_thread, /* attr = */ NULL,
				backend_handler, /* arg = */ NULL)) {
		char buf[1024];
		sdb_log(SDB_LOG_ERR, "Failed to create backend handler thread: %s",
				sdb_strerror(errno, buf, sizeof(buf)));

		plugin_main_loop.do_loop = 0;
	}
	else {
		size_t i;

		sdb_fe_socket_t *sock = sdb_fe_sock_create();
		for (i = 0; i < listen_addresses_num; ++i)
			if (sdb_fe_sock_add_listener(sock, listen_addresses[i]))
				break;

		/* break on error */
		if (i >= listen_addresses_num)
			sdb_fe_sock_listen_and_serve(sock, &frontend_main_loop);

		sdb_log(SDB_LOG_INFO, "Waiting for backend thread to terminate");
		plugin_main_loop.do_loop = 0;
		pthread_kill(backend_thread, SIGINT);
		pthread_join(backend_thread, NULL);
		sdb_fe_sock_destroy(sock);
	}

	sdb_log(SDB_LOG_INFO, "Shutting down SysDB daemon "SDB_VERSION_STRING
			SDB_VERSION_EXTRA" (pid %i)", (int)getpid());
	return 0;
} /* main */

/* vim: set tw=78 sw=4 ts=4 noexpandtab : */

